thoth-api/src/model/location/crud.rs

```diff
@@ -126,7 +126,101 @@ impl Crud for Location {
         crate::model::publication::Publication::from_id(db, &self.publication_id)?.publisher_id(db)
     }
 
-    crud_methods!(location::table, location::dsl::location);
+    // `crud_methods!` cannot be used for update(), because we need to execute multiple statements
+    // in the same transaction for changing a non-canonical location to canonical.
+    // These functions recreate the `crud_methods!` logic.
+    fn from_id(db: &crate::db::PgPool, entity_id: &Uuid) -> ThothResult<Self> {
+        let mut connection = db.get()?;
+        match location::table
+            .find(entity_id)
+            .get_result::<Self>(&mut connection)
+        {
+            Ok(t) => Ok(t),
+            Err(e) => Err(ThothError::from(e)),
+        }
+    }
+
+    fn create(db: &crate::db::PgPool, data: &NewLocation) -> ThothResult<Self> {
+        let mut connection = db.get()?;
+
+        connection.transaction(|connection| {
+            diesel::insert_into(location::table)
+                .values(data)
+                .get_result::<Self>(connection)
+                .map_err(|e| e.into())
+        })
+    }
+
+    fn update(
+        &self,
+        db: &crate::db::PgPool,
+        data: &PatchLocation,
+        account_id: &Uuid,
+    ) -> ThothResult<Self> {
+        // if changes to a location don't change its canonical or non-canonical status, only update that location.
+        if data.canonical == self.canonical {
+            let mut connection = db.get()?;
+            connection.transaction(|connection| {
+                match diesel::update(location::table.find(&self.location_id))
+                    .set(data)
+                    .get_result::<Self>(connection)
+                {
+                    // On success, create a new history table entry.
```

Comment:
To avoid repeating this logic across two branches, could store the `diesel::update` result(s) to a variable and then `match` on it at the very end of the function.

Reply:
great idea, thanks!